lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1827,28 +1859,64 @@ class BuildPipelineGroups : public FuncOpPartialLoweringPattern {
                                  PatternRewriter &rewriter) const {
     // Collect pipeline registers for stage.
     auto pipelineRegisters = getComponentState().getPipelineRegs(stage);
-
-    // Collect group names for the prologue or epilogue.
-    SmallVector<StringAttr> prologueGroups;
-    SmallVector<StringAttr> epilogueGroups;
-
     // Get the number of pipeline stages in the stages block, excluding the
     // terminator. The verifier guarantees there is at least one stage followed
     // by a terminator.
     size_t numStages = whileOp.getStagesBlock().getOperations().size() - 1;
     assert(numStages > 0);
 
-    // For each registered stage result value.
-    for (auto &operand :
-         stage.getBodyBlock().getTerminator()->getOpOperands()) {
+    // Collect group names for the prologue or epilogue.
+    SmallVector<StringAttr> prologueGroups, epilogueGroups;
+
+    auto updatePrologueAndEpilogue = [&](calyx::GroupOp group) {
+      // Mark the group for scheduling in the pipeline's block.
+      getComponentState().addBlockScheduleable(stage->getBlock(), group);
+
+      // Add the group to the prologue or epilogue for this stage as
+      // necessary. The goal is to fill the pipeline so it will be in steady
+      // state after the prologue, and drain the pipeline from steady state in
+      // the epilogue. Every stage but the last should have its groups in the
+      // prologue, and every stage but the first should have its groups in the
+      // epilogue.
+      unsigned stageNumber = stage.getStageNumber();
+      if (stageNumber < numStages - 1)
+        prologueGroups.push_back(group.sym_nameAttr());
+      if (stageNumber > 0)
+        epilogueGroups.push_back(group.sym_nameAttr());
+    };
+
+    MutableArrayRef<OpOperand> operands =
+        stage.getBodyBlock().getTerminator()->getOpOperands();
+    bool isStageWithNoPipelinedValues =
+        operands.empty() && !stage.getBodyBlock().empty();
+    if (isStageWithNoPipelinedValues) {
+      // Covers the case where there are no values that need to be passed
+      // through to the next stage, e.g., some intermediary store.
+      auto it = llvm::find_if(stage.getBodyBlock(), [&](auto &op) {
+        Optional<calyx::GroupOp> group =
+            getComponentState().getNonPipelinedGroupFrom<calyx::GroupOp>(&op);
+
+        if (!group.hasValue())
+          return false;
+
+        updatePrologueAndEpilogue(*group);
```

Comment:
This is getting a little wacky... this `updatePrologueAndEpilogue` has side effects, and is used inside the callback to `llvm::find_if`. I think that is an abuse of `find_if`. Please don't use `find_if` in this way, IMO this is being too clever. If you just need to loop through the body, do something like:

```
bool found = false;
for (...) ...
if (found)
   ...
```

Reply:
I see how transforming a "querying-like algorithm" into a state-modifying one is probably frowned upon, but I'm hard-pressed to believe that having some extra mutable state with a for-loop is a better choice.

Another suggestion would be to write some wrapper of `find_if`, e.g., `find_and_apply_if`. WDYT?

Reply:
Well, this is _local_ mutable state, which in my mind makes this more simple to grok when we're debugging it in 6 months. This pattern isn't unheard of: https://github.com/llvm/circt/search?q=%22bool+found%22&type=code.

I guess I am coming from a mentality of: we are writing C++, it's an icky language, and it's better to write dead simple C code (and more of it) than fancy code (and less of it). The llvm support stuff can help make it more readable in some cases, but this crossed a line I had for readability versus keeping things simple for my mental C interpreter. That said, if you want to use the callback style here, and can come up with a way to make that both terse and simple to understand in the future, I'm all for it.

Reply:
I'm not sure what you mean by icky. I think the best path forward is always writing safer code. One could say that a the "mental interpreter" is a function of what you're accustomed to seeing. This is a tale as old as time; std::variant versus union, malloc versus unique_ptr, etc. If we don't allow modern C++ features to appear in code, then yes, they will always be difficult to parse.